azure/src/integration/java/org/apache/iceberg/azure/adlsv2/ADLSFileIOTest.java

```diff
@@ -190,4 +200,188 @@ public void testSerialization(TestHelpers.RoundTripSerializer<FileIO> roundTripS
 
     assertThat(testFileIO.properties()).isEqualTo(roundTripSerializedFileIO.properties());
   }
+
+  @Test
+  public void noStorageCredentialConfigured() {
+    ADLSFileIO fileIO = new ADLSFileIO();
+    fileIO.setCredentials(ImmutableList.of());
+    fileIO.initialize(
+        ImmutableMap.of(
+            sasTokenForAccount("account1"), "sasTokenFromProperties1",
+            sasTokenForAccount("account2"), "sasTokenFromProperties2"));
+
+    assertThat(fileIO.clientForStoragePath(abfsPath("my-container1", "account1", "/path/to/file")))
+        .isSameAs(
+            fileIO.clientForStoragePath(abfsPath("my-container2", "account2", "/path/to/file")))
+        .isSameAs(fileIO.clientForStoragePath(wasbPath("container", "account", "/path/to/file")))
+        .isSameAs(fileIO.clientForStoragePath(abfsPath("random", "account", "/path/to/file")));
+
+    assertThat(
+            fileIO
+                .clientForStoragePath(abfsPath("my-container1", "account1", "/path/to/file"))
+                .azureProperties())
+        .extracting("adlsSasTokens")
+        .asInstanceOf(InstanceOfAssertFactories.map(String.class, String.class))
+        .containsEntry(accountHost("account1"), "sasTokenFromProperties1")
+        .containsEntry(accountHost("account2"), "sasTokenFromProperties2");
+  }
+
+  @Test
+  public void singleStorageCredentialConfigured() {
+    StorageCredential adlsCredential =
+        StorageCredential.create(
+            abfsPath("custom-container", "account1", ""),
+            ImmutableMap.of(sasTokenForAccount("account1"), "sasTokenFromCredential"));
+    ADLSFileIO fileIO = new ADLSFileIO();
+    fileIO.setCredentials(ImmutableList.of(adlsCredential));
+    fileIO.initialize(
+        ImmutableMap.of(
+            sasTokenForAccount("account1"), "sasTokenFromProperties1",
+            sasTokenForAccount("account2"), "sasTokenFromProperties2"));
+
+    // verify that the generic ADLS client is used if the storage prefix doesn't match the prefixes
+    // in the storage credentials
+    assertThat(
+            fileIO.clientForStoragePath(abfsPath("custom-container", "account1", "/path/to/file")))
+        .isNotSameAs(
+            fileIO.clientForStoragePath(abfsPath("my-container", "account1", "/path/to/file")));
+
+    assertThat(fileIO.clientForStoragePath(abfsPath("my-container", "account1", "/path/to/file")))
+        .isSameAs(
+            fileIO.clientForStoragePath(abfsPath("my-container2", "account2", "/path/to/file")));
+
```

Comment:
Same as what I mentioned somewhere else in this test, could we make sure some of these `abfsPath(...)` are moved to separate values and can we make sure these tests are strong as possible beyond just comparing if it's the same as another path.